spec/ics-007-tendermint-client/README.md

```diff
@@ -114,14 +126,20 @@ Tendermint client validity checking uses the bisection algorithm described in th
 function checkValidityAndUpdateState(
   clientState: ClientState,
   header: Header) {
-    // assert that header is newer than any we know
-    assert(header.height < clientState.latestHeight)
+    // assert trusting period has not yet passed
+    assert(currentTimestamp() - clientState.latestTimestamp < trustingPeriod)
+    // assert header timestamp is not in the future (& transitively that is not past the trusting period)
+    assert(header.timestamp <= currentTimestamp())
```

Comment:
Don't have an intuitive sense of why these 2 checks are necessary.

The first check is ensuring that we don't allow updates to happen if the last update to the connection chain happened before our chain's current time minus the trustingPeriod. What is the purpose of this check? Seems like it will force connections to update within a specific frequency, however i don't see the danger of allowing connections to update as slow as they wish (perhaps a connection with very rare messages sent through it, may choose to update lazily?). It also seems like if this assertion ever fails, then this connection is no longer updatable at all. Is this the desired behavior?

The second check prevents a connecting chain from updating with a header if its timestamp happens to be greater than the current timestamp. This doesn't seem to be an issue to have this check, but i don't understand the attack it is preventing.

Don't think these are necessarily issues to be resolved, but would like an explanation of why these checks are necessary

Reply:
> The first check is ensuring that we don't allow updates to happen if the last update to the connection chain happened before our chain's current time minus the trustingPeriod. What is the purpose of this check? Seems like it will force connections to update within a specific frequency, however i don't see the danger of allowing connections to update as slow as they wish (perhaps a connection with very rare messages sent through it, may choose to update lazily?). It also seems like if this assertion ever fails, then this connection is no longer updatable at all. Is this the desired behavior?

This is the client update logic - the "trusting period" is just the "unbonding period", less some buffer of time intended to allow for the submission of evidence if misbehaviour happens.

> The second check prevents a connecting chain from updating with a header if its timestamp happens to be greater than the current timestamp. This doesn't seem to be an issue to have this check, but i don't understand the attack it is preventing.

Because it might no longer fall within the trusting period (unbonding period).